### ISSUE TITLE
Another file where flatten-annotations loses the data

### ISSUE THREAD

**Issue by jroldroyd** [qpdf/qpdf#949]
When filling out and printing the IRS Form 941 PDF.  After entering data and printing the form using cups (which uses the QPDF libs), the page prints but the form annotations are lost.

Running "qpdf --flatten-annotations=all" on the file also results in blanks where the form data was.

The Input and output files and more details are on the FreeBSD PR I filed here:

[https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=270832](url)

**Comment by jberkenbilt** [comment]
This is the same problem as https://github.com/OpenPrinting/libcupsfilters/issues/28

**Comment by jberkenbilt** [comment]
I have reproduced this and have a one-line fix. I need to study this to figure out why this has never been found before so I can figure out why I originally coded it the way I did. Fixing doesn't cause any test failures, so I may have just missed case. It's odd that this was never found before but has been found a few times just in the last month.

**Comment by jberkenbilt** [comment]
I see what's going on. The file actually has an error, but I can work around it.

**Comment by akorobkin** [comment]
> I figured out what's going on here. The PDF file in question is using the /AS key (appearance state) incorrectly, but qpdf can easily work around it.

Jay, 

is there any advice you can give to the developers of the PDF software that filled out that form? What do they need to change to make sure the form is filled correctly? I can try to file the bugs to respective upstream projects.

**Comment by jroldroyd** [comment]
I have applied the patch to the FreeBSD code and can confirm that the file now both displays okay after --flatten-annotations and also prints okay.

Thanks very much for the patch, Jay.

**Comment by jberkenbilt** [comment]
I'm away from my computer today and tomorrow but intend to answer your question about advice to the other devs, so I'm reopening as a reminder. Basically they need to refresh themselves on what `/AS` is used for when I'm back I will reference the specific sections from the PDF spec and resolve the unrelated 32 bit failures in CI so I can merge my fix. I will probably cut an 11.4 release soon with this and a few other minor things.

**Comment by jberkenbilt** [comment]
@jroldroyd In the latest version of the PDF SPEC, ISO 32000-2:2020, which is now [available at no cost](https://pdfa.org/resource/iso-32000-pdf/), you can find Table 166 in section 12.5.2 which describes the entries in the annotation dictionary. The `/AS` key is described as

> (Required if the appearance dictionary **AP** contains one or more subdictionaries; PDF 1.2) The annotation’s appearance state, which selects the applicable appearance stream from an appearance subdictionary (see 12.5.5, "Appearance streams").

The appearance dictionary's top-level keys are defined in Table 170 in Section 12.5.5:

| Key | Type | Value |
| --- | --- | --- | --- | --- |
| N | stream or dictionary(Required) | The annotation’s normal appearance. |
| R | stream or dictionary(Optional) | The annotation’s rollover appearance. |
| D | stream or dictionary(Optional) | The annotation’s down appearance. |

(I don't know why that is not rendering as a table, but hopefully it's clear.)

When would `/N` be a stream, and when would it be a stream and when would it be a dictionary? The answer is that it would be a stream for simple annotation types like text and a dictionary for annotation types that can have one of many _appearance states_ such as radio buttons or checkboxes.

The incorrect PDF files in this issue set `/AS` to `/N`, which is not correct. The top-level keys (`/N`, `/R`, or `/D`) all point to appearance streams or dictionaries, and the one that is used is dependent on what the user is doing. It doesn't make sense to encode that into the file. The `/AS` key should be omitted for text fields or other annotations that don't have multiple appearance states. For an example of where you would use `/AS`, see 12.7.5.2.3 **Check boxes**.

So, bottom line, whoever created that PDF didn't understand the purpose of `/AS`. It should never be set to the value of `/N` -- setting it to that value is meaningless.

**Comment by tillkamppeter** [comment]
Tested the fix for the case of https://github.com/OpenPrinting/libcupsfilters/issues/28 and this fix indeed solves the problem. Thanks a lot @jberkenbilt !

**Comment by jberkenbilt** [comment]
Great to hear. @tillkamppeter As you are surely aware, this fix was included in 11.4.0 which was released yesterday.
